ScheduledJob controller proposal

soltysh opened the pull request:
Diverged from #11746.

// cc @smarterclayton @timothysc @pmorie @bgrant0607 @davidopp @nwendling @derekwaynecarr @erictune @mikedanese @thockin

soltysh committed:
ScheduledJob controller proposal

timothysc commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,132 @@
+# ScheduledJob Controller
```
Comment still holds on name

smarterclayton replied:
The problem with "Cron" is that is technically focused (a particular linux
tool) but is opaque to higher level users who come from different
backgrounds.

Even the Chronos GitHub page describes it with "It allows you to schedule
your jobs using ISO8601 repeating interval notation, which enables more
flexibility in job scheduling".  RepeatingJob or CalendarJob or
RecurringJob all seem less clear to an outside user (someone who is not an
expert at these systems) than ScheduledJob.

On Fri, Jul 31, 2015 at 12:41 PM, Timothy St. Clair <
notifications@github.com> wrote:

> In docs/proposals/scheduledjob.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11980#discussion_r35991090>
> :
>
> > @@ -0,0 +1,132 @@
> > +# ScheduledJob Controller
>
> Comment still holds on name
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11980/files#r35991090>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

timothysc replied:
Proposal was around (diurnal, periodic, etc.) There may be a couple more.

smarterclayton replied:
Periodic suffers some of the same usability concerns.  I do think API
objects should be precise in their meaning, but designed for end users, not
systems researchers.

On Fri, Jul 31, 2015 at 1:02 PM, Timothy St. Clair <notifications@github.com
> wrote:

> In docs/proposals/scheduledjob.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11980#discussion_r35992880>
> :
>
> > @@ -0,0 +1,132 @@
> > +# ScheduledJob Controller
>
> Proposal was around (diurnal, periodic, etc.) There may be a couple more.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11980/files#r35992880>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

bgrant0607 replied:
I'm fine with ScheduledJob. It's similar to ScheduledExecutorService in Java.

bgrant0607 replied:
Could also call it CalendaredJob

erictune commented:
When you implement this please put in expapi.  (#12001).

soltysh commented:
@erictune gotcha.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,132 @@
+type ScheduledJobStatus struct {
+
+    // PendingExecutions are the job runs pending for execution
+    PendingExecutions []Job
```
Much of the same feedback applies to this API as to Job.

soltysh replied:
Yeah, I'll address them as soon as we have consensus on Job API, which I'm hoping will happen withing next couple of days :)

pmorie replied:
My feedback re: Job API and overall status applies here too.

soltysh commented:
@bgrant0607 @erictune I've updated the proposal, cutting the status structure significantly. I've only left there a list of currently executing jobs, which is needed to track these executions and fulfill the defined ConcurrentPolicy.

soltysh commented:
@derekwaynecarr @smarterclayton ptal as well

mikedanese commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,161 @@
+type ScheduledJobStatus struct {
+
+    // Jobs are the currently running instances of a ScheduledJob.
+    Jobs []Job
```
Why not a selector query? (you'd have to add a selector to ScheduledJobSpec)

soltysh replied:
How would you differentiate concurrent runs? We'd have to generate some random uuid for that selector in that case.

mikedanese replied:
The selector should select all Jobs created by  the ScheduledJob. We can sort them by CreationTimestamp. What do you mean differentiate? How does having an array help you differentiate?

soltysh replied:
Yeah, I haven't thought about sorting them by CreationTimestamp, which generally speaking allows me to distinguish (that's the word I've meant to use, sorry :/) different ScheduledJob executions/runs needed to properly handle ConcurrentPolicy.

bgrant0607 replied:
Sorting by CreationTimestamp seems exactly what the user would want, since ScheduledJobs are launched based on time.

Please no array.

soltysh replied:
Array removed, added selector to `ScheduledJobSpec`.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,161 @@
+    Jobs []Job
+
+    // Successful tracks the overall amount of successful completions of this job.
+    Successful int
```
If these accumulate indefinitely, they should probably be int64.

soltysh replied:
Done.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,161 @@
+
+    // SkipNextConcurrent forbids concurrent runs, skipping next run if previous
+    // hasn't finished yet.
+    SkipNextConcurrent ConcurrentPolicy = "SkipNextConcurrent"
```
You don't need to include "Concurrent" in all the values. See, for instance, RestartPolicy.

Also, I'd call this ConcurrencyPolicy.

soltysh replied:
Done.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,161 @@
+    Schedule string
+
+    // ConcurrentPolicy specifies how to treat concurrent executions of a Job.
+    ConcurrentPolicy ConcurrentPolicy
```
Is there a default value?

soltysh replied:
Yes, it's AllowConcurrent. I've put it on the type declaration, similarly like it's done with RestartPolicy.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,161 @@
+    ConcurrentPolicy ConcurrentPolicy
+
+    // BlockOnFailure suspends scheduling of next job runs after a failed one.
+    BlockOnFailure bool
```
Is there a default value?

soltysh replied:
It's false, I'll update the description.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,161 @@
+* repeatedly at a specified point in time.
+
+There is already a discussion regarding that particular subject:
+* Distributed CRON jobs [#2156](https://github.com/GoogleCloudPlatform/kubernetes/issues/2156)
```
Repo has moved. Use the redirector:
http://issues.k8s.io/2156

soltysh replied:
Fixed.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,161 @@
+type ScheduledJobSpec struct {
+
+    // JobSpec is a structure defining the expected behavior of a job.
+    JobSpec JobSpec
```
For comparison, see the Deployment proposal: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/deployment.md

soltysh replied:
After talking to @smarterclayton I've renamed JobSpec to be pointer to JobSpec.

soltysh commented:
@mikedanese @bgrant0607 addressed your comments, appreciate another round of reviews.

mikedanese commented:
Let's continue this review targeting v1.2

sdminonne commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+    // - 2015-07-21T14:00:00Z - represents date and time in UTC
+    // - R/2015-07-21T14:00:00Z/P1D - represents endlessly repeating interval (1 day), starting from given date
+    Schedule string
+
```
@soltysh my understanding here is that ISO8601 format it's going to be a necessary pain but as far as I understood `string` are not mergeable and to me schedule spec looks something we may want to update quite often. Should we consider a more structured representation instead?

soltysh replied:
My understanding is that you won't be merging schedule, but rather replacing an old one with a new one. Although ISO8601 format seems clunky at first, it's very readable and easily customizable. Using any kind of structured representation would introduce unnecessary complication imho.

sdminonne replied:
I was not discussing ISO8601 here but the string representation.
I'm not sure I fully understand all the implications of replacing schedules. Do you plan to delete/create? In any cases may you add something about it in the doc (except if I missed something)?

soltysh replied:
My general idea (which as you point out needs clear statement somewhere in this proposal) is that any changes you'll do to the scheduledjob will reflect in every new run, it does not affect nor delete existing executions. Eg. I have a job running every 15 mins, it's 1.45, so one is currently running, and I change the schedule to be running every hour, that means the existing job will finish what it started and next one will start at 2, then at 3 etc. @sdminonne does it make sense?

sdminonne replied:
@soltysh absolutely but when you say _I change the schedule to be running every hour_ are you updating the scheduleJob spec or are you creating another one? My understanding is that we should simply updating the current scheduleJob. But to do this we need the `iso8601 spec` to me mergeable. Am I missing something?

mikedanese replied:
I don't see much utility in partially updating a schedule over replacing a schedule.

soltysh replied:
Right, those where my feelings as well, though I know nothing about mergeable strings and need to read about it a bit. But again, when updating you'll just modify the current schedule so that it matches your new expectations and submit the new one which will replace the old one.

bgrant0607 replied:
Maybe an example would help.

For instance, one might change `R/2015-07-21T14:00:00Z/P1D` to `R/2015-07-26T20:00:00Z/P1D`. 

@sdminonne Do you want to be able to independently change the start date, start time, and periodicity?

sdminonne replied:
@bgrant0607, doesn't matter replacing completely the scheduler is good enough for us. Thanks

mikedanese commented:
@soltysh are you going to work on this for v1.2?

soltysh commented:
@mikedanese yes sir, I'm waiting for reviews  but I guess those will happen when 1.1 will be closed. You're welcome to comment at any time, though.

mikedanese commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+
+    // BlockOnFailure suspends scheduling of next job runs after a failed one,
+    // by default it is false.
+    BlockOnFailure bool
```
Isn't this just a type of concurrency policy?

mikedanese replied:
Misread the doc. What is a failed run?

soltysh replied:
Hmmm... with current Jobs API I'll probably have to introduce some kind of succeeded/failed pod rate that will allow deciding what is a failed run, WDYT?

mikedanese replied:
Sounds good. Should we remove from proposal until that is implemented?

soltysh replied:
I'm not sure if that should be part of a job, though. Initially I was thinking this would be only applicable here, but thinking about it more it would benefit jobs actually. @erictune wdyt? This should be pretty easy to implement actually.

soltysh replied:
@bgrant0607 ptal as well, regarding job API changes, or should I create a separate proposal for that?

bgrant0607 replied:
Is this thread about Concurrency and BlockOnFailure? Those don't make sense for Job.

soltysh replied:
It's about having an option (in jobs) which allows defining when a job is considered failed. Having thought about it more, though I think that it's client role to decide when a job succeeded or failed. In this case it's the ScheduledJob that is Job's client that needs that information to decided about job.

soltysh commented:
@mikedanese are there any other objections/questions for this proposal, I'd love to see this moving forward, so that I can put my hands on the implementation bits :wink:

mikedanese commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+    Successful int64
+
+    // Unsuccessful tracks the overall amount of failures of this job.
+    Unsuccessful int64
```
Jobs can't fail.

soltysh replied:
That will be dependent on failure rate on jobs/scheduled jobs, we're discussing https://github.com/kubernetes/kubernetes/pull/11980#discussion_r44013029. For now I'll remove it then.

mikedanese commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+
+    // SkipNextConcurrent forbids concurrent runs, skipping next run if previous
+    // hasn't finished yet.
+    SkipNextConcurrent ConcurrencyPolicy = "SkipNext"
```
This is confusing because with SkipNextConcurrent, the object in the statement "skip the next job" is the job that is already running and with CancelPreviousConcurrent, the object in the statement "skip the previous job" is the job that is up to be run. Side note, with AllowConcurrent the subject in the statement "allow this job" is the job that is up to be run like CancelPreviousConcurrent.

soltysh replied:
I see your point, but atm I can't think of anything better to represent my idea, which I'm hoping is clear for you.

soltysh replied:
How about: AllowConcurrent, ForbidConcurrent and ReplaceConcurrent? These 3 should target what you're asking for.

sdminonne replied:
@soltysh @mikedanese for recurring job would accept to add a simple state like

```go
type ScheduledJobStatus string

const {
    RunningStatus ScheduledJobStatus  = "Running"
    SuspendedStatus ScheduledJobStatus = "Suspended"
}
```
defaulted to `Running`. Obviously this will add the ability to suspend the scheduling for example via `edit` command.

sdminonne replied:
@soltysh another topic is updating of the scheduler for repeating scheduledJob we may want to implement something like a rolling update to replace the scheduler if we [cannot update](https://github.com/kubernetes/kubernetes/pull/11980#discussion_r40263306) the resource

soltysh replied:
@sdminonne the status is worth getting into the ScheduledJobStatus structure.

> if we cannot update the resource

We can, we won't forbid that. We've stated there is no partial updates to schedule, just full replacement.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+* repeatedly at a specified point in time.
+
+There is already a discussion regarding that particular subject:
+* Distributed CRON jobs [#2156](https://github.com/kubernetes/kubernetes/issues/2156)
```
Use redirector: issues.k8s.io/2156

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+
+## Motivation
+
+ScheduledJobs are needed for performing all time related actions, namely backups,
```
time-related
English is right-associative.

bgrant0607 commented:
@soltysh @mikedanese @erictune Has work resumed on this yet? I assume it's needed for 1.2?

bgrant0607 commented:
When work does resume:
* bring API in line with conventions
* apply changes/lessons from Job (e.g., generalized label selector)

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+
+    // Template is the object that describes the job that will be created when
+    // executing a ScheduledJob.
+    Template *JobSpec
```
JobSpec doesn't include metadata, so it doesn't include labels, so it's not possible to ensure that the jobs match the specified selector. Take a look at ReplicationController and Deployment.

soltysh commented:
@bgrant0607 yes, I'm currently working on v2 of this proposal. Until now I was waiting for initial review for which I'm very grateful. I'm hoping to have the update by the end of this week.

pmorie commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+* once at a specified point in time,
+* repeatedly at a specified point in time.
+
+There is already a discussion regarding that particular subject:
```
nit: s/that particular subject/this subject/

pmorie commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+## Motivation
+
+ScheduledJobs are needed for performing all time related actions, namely backups,
+report generation and alike.  Each of these tasks should be allowed to perform
```
nit: s/alike/the like/

pmorie replied:
s/perform/run/

pmorie commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+
+## Implementation
+
+ScheduledJob controller relies heavily on the [Job Controller API](job.md)
```
nit: `The ScheduledJob...`

pmorie commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+
+    // Status is a structure describing current status of a job.
+    Status ScheduledJobStatus
+}
```
I would go ahead and add json tags into this struct, there's no reason not to fully define API objects in the proposal (especially ideally they would be the exact same as what went into the code).

soltysh replied:
I didn't do it up until now not to clutter the proposal with too much of implementation details. The comments are exactly like the ones in the actual code.

soltysh replied:
Maybe except for the places where I'm just showing the fact of existence of a structure, eg. `ScheduledJobCondition` in this proposal (to be added in v2), since they follow exact patterns as similar structs elsewhere `JobCondition`, `PodCondition`, etc.

pmorie commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,163 @@
+    ListMeta
+
+    Items []ScheduledJob
+}
```
Same comment applies here: json tags / comments.

pmorie commented on docs/proposals/scheduledjob.md:
````diff
@@ -0,0 +1,163 @@
+}
+```
+
+`ScheduledJobSpec` structure is defined to contain all the information how the actual
````
Nit: `The ScheduledJobSpec` structure contains all the...`

pmorie commented on docs/proposals/scheduledjob.md:
````diff
@@ -0,0 +1,163 @@
+)
+```
+
+`ScheduledJobStatus` structure is defined to contain information about scheduled
````
Nit: start sentence w/ `The ScheduledJobStatus...`

pmorie commented on docs/proposals/scheduledjob.md:
````diff
@@ -0,0 +1,163 @@
+    Unsuccessful int64
+}
+```
+
````
Add examples here?

soltysh replied:
I think they are already covered by use cases, I don't see any point in repeating those.

soltysh commented:
I've updated the proposal, @erictune @bgrant0607 @mikedanese @pmorie ptal

@mikedanese given #17244 and #17242 I've decided to leave the failed-related fields.

davidopp commented:
Sorry for jumping in without reading the full history, but maybe ScheduledJobSpec should just point to a JobSpec and then add additional information? @bgrant0607 has in the past suggested some kind of "inert" objects (I believe that's what he called them), which are kind of like templates, I guess; the JobSpec pointed to by the ScheduledJobSpec would be inert.

Alternatively (perhaps better), how about refactoring the common parts out of JobSpec and ScheduledJob into a new object (a JobTemplate, or something like that), and have them both point to that?

I think the current structure is bound to let the definitions of Job and ScheduledJob get out of sync, which I think is undesirable.

c4milo commented:
Does this allow defining job dependencies like in Chronos?

sdminonne commented:
@c4milo no, it doesn't. I'm working on a proposal to create jobs sequence. It should land in few days (hopefully not weeks), I'll ref you for a feedback (I'm not a Chronos expert since we use another system).

mikedanese commented:
DAG workflow execution would be part of Job API and is not specific to ScheduledJob API

soltysh commented:
@davidopp actually it used to have JobSpec (see the first commit), but per @bgrant0607 suggestion I've extracted those fields into ScheduledJobSpec, unless I misunderstood something. Currently it's similar to how Deployments define RC.

sdminonne commented:
@mikedanese is there a draft for DAG workflow? Or may you post here the GH ref?

davidopp commented:
> @davidopp actually it used to have JobSpec (see the first commit), but per @bgrant0607 suggestion I've extracted those fields into ScheduledJobSpec, unless I misunderstood something. Currently it's similar to how Deployments define RC.

I talked to @bgrant0607 about this and I think you may have read too much into his suggestion to follow the pattern from Deployment. He agreed that you should include the JobSpec in the ScheduledJobSpec. He also suggested moving up the fields that are most important, and leaving out BlockOnFailure for now. So it would look something like

```
ScheduledJobSpec {
    Schedule string
    ConcurrencyPolicy ConcurrencyPolicy
    JobTemplate JobSpec
    UniqueLabelKey string
}
```

soltysh commented:
Makes sense. I'll update my proposal tomorrow morning.
On Nov 16, 2015 11:16 PM, "David Oppenheimer" <notifications@github.com>
wrote:

> @davidopp <https://github.com/davidopp> actually it used to have JobSpec
> (see the first commit), but per @bgrant0607
> <https://github.com/bgrant0607> suggestion I've extracted those fields
> into ScheduledJobSpec, unless I misunderstood something. Currently it's
> similar to how Deployments define RC.
>
> I talked to @bgrant0607 <https://github.com/bgrant0607> about this and I
> think you may have read too much into his suggestion to follow the pattern
> from Deployment. He agreed that you should include the JobSpec in the
> ScheduledJobSpec. He also suggested moving up the fields that are most
> important, and leaving out BlockOnFailure for now. So it would look
> something like
>
> ScheduledJobSpec {
>     Schedule string
>     ConcurrencyPolicy ConcurrencyPolicy
>     JobTemplate JobSpec
>     UniqueLabelKey string
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/11980#issuecomment-157188820>
> .
>

soltysh commented:
Sorry for the slip, here's the updated proposal. There's one more thing I'm currently thinking about. Athough we've decided to remove `BlockOnFailure` for now, I prefer to leave the conditions in status, especially the one related to suspensions. I still would like to see an option to suspend a scheduled job, currently I'm thinking about a new kubectl command. Sort of pause/suspend, name to be decided. wdyt?

timothysc commented:
@davidopp with multiple schedulers in play, do we need 1st classed object support / controllers for this.  Or explicitly defer to other scheduler and leverage annotations?

davidopp commented:
@timothysc My understanding of the design is
* there will be a new ScheduledJob object (described in this PR)
* we will write a new controller to handle ScheduledJob; at the appropriate times, it will create a Job object, which will then be handled by the existing Job controller
* if you want a different scheduler to handle pods created by Job/ScheduledJob vs. pods created by ReplicationController, then you write an initializer that somehow distinguishes them and writes an annotation with the name of the appropriate scheduler.

Having a back-pointer from pod to its controller is part of "controllers v2" (#14961) and not yet implemented, so the initializer would need some other way to distinguish which controller created the pod... I'm not aware of any way yet today (but maybe I'm forgetting something).

Does that answer your question, or did I misinterpret?

timothysc commented:
@davidopp kind... of... 

This seems like a point in space-time thing.  I would bet that folks are going to create their own workflow and time based scheduler&controllers as evolution takes its course.

davidopp commented:
@timothysc Oh, now I understand your question -- it was "do we really need this" ?

I guess I would say, if someone wants to make something better in the very near future, maybe we shouldn't bother. But nobody has stepped forward to build something else, and we have received requests for something cron-like, so this seems reasonable for now...

soltysh commented:
@erictune @davidopp @mikedanese @sdminonne guys can I have another round of reviews, pls. I'd like to push this forward :)

davidopp commented:
Can you explain what causes a job to be "suspended" (the ScheduledJobSuspended condition)?
Also I don't think the LastProbeTime field is necessary there.

soltysh commented:
> Can you explain what causes a job to be "suspended" (the ScheduledJobSuspended condition)?

My idea is to manually suspend a job, when needed. This will need a new action in the API. I should have probably added that to the proposal; I'll add it to the next revision of the proposal if you don't mind. I guess it'll need new action in kubectl as well to support that as well.

> Also I don't think the LastProbeTime field is necessary there.

I'm following our API conventions, described [here](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#typical-status-properties).

mikedanese commented:
API conventions say a subset so it's not required, and we should only use heartbeats on conditions when we absolutely need it to function as it kills scalability right now. Also if you say that the user suspends a job, then it should be a field in Spec not Status.

soltysh commented:
I'll remove the LastProbeTime. I'm wondering though, if it should be removed from Jobs as well?
I'll also add appropriate field in Spec, a boolean flag allowing to suspend a scheduled job. I'll update the proposal with a new version tomorrow morning.

mikedanese commented:
> I'm wondering though, if it should be removed from Jobs as well?

yes it should be.

> I'll also add appropriate field in Spec, a boolean flag allowing to suspend a scheduled job. I'll update the proposal with a new version tomorrow morning.

Acknowledgement of a controller that it has seen a field in spec is done using Generation and ObservedGeneration

erictune commented:
A single ISO8601 does not allow specifying monday thru friday runs.  There is an open issue for chronos to support this: https://github.com/mesos/chronos/issues/194

Should there be a list of periods, so you can express monday thru friday with a list of 5 weekly periods?

Also, what about hourly throughout business hours?

erictune commented:
Chronos has a "scheduleTimeZone" which overrides the "schedule".  Can you figure out what that is for and if this needs it too?

erictune commented:
Say the schedule is R/2015-12-01T10:00:00Z/P1D.
Say that it is currently 2015-12-01T11:00:01Z ( one hour past the first run time). 

Say the controller has been down for an unknown amount of time.  Now it is restarted and is working right.

How does it know whether it should start the job, versus if it has previously started it and it completed successfully?

Basically, what state does it need to store about what it has previously done, and does any of that need to be in the ScheduledJob object?  

Also, Chronos has an "epsilon" field, which controls how long after the intended start time that the controller will start a job.    Do we need that?

erictune commented:
When `AllowConcurrent` is used, what make the two jobs have different pod selectors.

bgrant0607 commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,184 @@
+
+    // Template is the object that describes the job that will be created when
+    // executing a ScheduledJob.
+    JobTemplate JobSpec
```
I think we may need a more general template here, to create arbitrary numbers of jobs and/or arbitrary types of resources.

soltysh replied:
What kind of general template you have in mind?

erictune replied:
If you support multiple JobTemplates, then it does make it more complex to report success or failure of the ScheduledJob, and for users to reason about partial success.  

Another way to run multiple Jobs, is to run a  [Workflow](https://github.com/kubernetes/kubernetes/pull/17787#issuecomment-162073407).

What if ScheduledJob supports running the user's choice of either a single Job or a single Workflow.

erictune replied:
Brian also mentioned offline that maybe you need to create `secrets` or `configData` each time a cron job runs.

However, I think you can create those by creating a Workflow, whose first Job creates any needed secrets or configData, and a dependent Job that consumes those resources.

soltysh replied:
Yeah, I've talked with @sdminonne about it and we came to a conclusion that we'll start working on top of current Job resource. Once he has the workflow and I have the scheduledjob in we'll start working on combining the two. Does this make sense?

erictune replied:
SGTM

soltysh commented:
@erictune:

> When AllowConcurrent is used, what makes the two jobs have different pod selectors?

UniqueLabelKey is to address that issue. It'll be additional selector (user provided or system generated if none is specified) that will allow distinguishing pods.

> How does it know whether it should start the job, versus if it has previously started it and it completed successfully?
Basically, what state does it need to store about what it has previously done, and does any of that need to be in the ScheduledJob object? 

I'm thinking of a last started job name (or better object ref) & time, which might address that issue. 

I'll have a closer look at the other issues you've mentioned:
- multiple schedules - an array of schedules seems reasonable, but we need to make sure the provided values don't overlap, imho;
- epsilon - hmm... with current implementation of jobs, and their reliability of the current scheduling mechanism in k8s I'm not sure we need anything like that;
- scheduleTimeZone - will check it;

soltysh commented:
Thanks for the valuable input! I'll start working on the next version of this proposal.

erictune commented:
Note that it is already possible to run a `Job` via `kubectl run`, when you specify `--restartPolicy=Never` or `OnFailure`.  

I suggest this proposal include a sketch of how to start a `ScheduledJob` using `kubectl run`.

Maybe something like:

```
$ kubectl run pi --image=perl --restart=OnFailure --runAt= 2015-07-21T14:00:00Z -- perl -Mbignum=bpi -wle 'print bpi(2000)' 
created "scheduledJob/pi"
```

soltysh commented:
@erictune ok, I'll start updating the proposal. Hopefully I should have a new version by the end of this week.

soltysh commented:
Sorry for the slip, here's the updated proposal, changes from previous version are:
* added `Suspend` flag in Spec, see [this comment](https://github.com/kubernetes/kubernetes/pull/11980#issuecomment-161114514);
  I'm not using Generation, since all the changes to spec applies only to subsequent executions, not currently running jobs.
* after giving the epsilon a second thought, I'm adding it to allow to specify a time within which a job can be run if it misses its schedule
* added `LastScheduleTime` in Status to keep information about when was the last successful execution
* added `Active` condition to denote the job is running
* added `kubectl run` extensions for running scheduled jobs 

> Chronos has a "scheduleTimeZone" which overrides the "schedule". Can you figure out what that is for and if this needs it too?

`scheduleTimeZone` is meant to override `schedule` time zone. In ISO8601 you specify TZ as an [offset](https://en.wikipedia.org/wiki/ISO_8601#Time_offsets_from_UTC) from UTC. In chronos they allow you to use TZ name from [TZ database](https://en.wikipedia.org/wiki/Tz_database).
I'm not sure we want to complicate the scheduling algorithm with this, especially that we're introducing multiple schedules. See further for other thoughs as well.

> Should there be a list of periods, so you can express monday thru friday with a list of 5 weekly periods?
> Also, what about hourly throughout business hours?

I'm starting to think if ISO8601 is such a good choice for representing the schedule. Having seen the problems chronos is dealing with, and the issues you've raised, wouldn't it be better/simpler to fall-back to using good old cron, which addresses above two cases with just a single line: `0 9-18 * * 1-5`? The problem with ISO8601 is that just to address the first issue (mon-fri) I'd need to extend the schedule field to be a list, which is not that bad. But to address the 2nd issue (hourly through  bussiness hours) I'd have to complicate the schedule to allow nested schedules to address above, is it really worth it?  I'm not sure about TZ though, but since all the times on the server are UTC based,  I'd say not to worry about it.

@erictune @mikedanese ptal

mikedanese commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,222 @@
+
+    // Optional duration in seconds the job will still run if it misses scheduled
+    // time for any reason.  Missed jobs executions will be counted as failed ones.
+    Epsilon *int64
```
There must be a better name for this.

soltysh replied:
You're right, following our [api conventions](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#naming-conventions) I'd say something along `MisfireTimeoutSeconds` or `MisscheduleTimeoutSeconds`, wdyt?

erictune replied:
AppEngine has a nice cron format that is similar to old fashioned cron, but more readable:
https://cloud.google.com/appengine/docs/python/config/cron?csw=1#Python_app_yaml_The_schedule_format

erictune replied:
`StartingDeadline` is the deadline the ScheduledJob controller has for starting the Job?

soltysh replied:
`StartingDeadlineSeconds` sgtm.

sdminonne commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,222 @@
+Below are the possible future extensions to the Job controller:
+* Be able to specify more general template in `.spec` field, to create arbitrary
+  types of resources. This relates to the work happening in [#7787](https://issues.k8s.io/17787)
+
```
@soltysh nit s/7787/17787/

soltysh replied:
Fixed.

erictune commented:
I noticed that Oozie has a cron-like feature, and that it originally used an ISO8601-like format, but then later added a cron-like format too because simple recurring events were not sufficient.  Worth a read:
http://oozie.apache.org/docs/4.2.0/CoordinatorFunctionalSpec.html#a4.4.3._Cron_syntax_in_coordinator_frequency

Based on that, I'm inclined to agree with you about supporting cron format.

erictune commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,222 @@
+
+    // Template is the object that describes the job that will be created when
+    // executing a ScheduledJob.
+    JobTemplate JobSpec
```
Can the JobTemplate have a `metadata.name`, or must it use `metadata.generateName`?  If the former, what happens in case of `Allow` concurrency?

soltysh replied:
`JobTemplate` does not have a `metadata.name` field, I was thinking about generating name, each one will be a scheduled job name and a increasing counter.

soltysh replied:
This will also address your 2nd part of the question when `Allow` concurrency is chosen.

erictune commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,222 @@
+    // Successful tracks the overall amount of successful completions of this job.
+    Successful int64
+
+    // Failed tracks the overall amount of failures of this job.
```
Do we need to distinguish between Failure Condition of the Job, and failure to start the job entirely (name collision, lack of permissions to create job, apiserver not available, etc)

soltysh replied:
I don't think so, at least for now I don't see the need for that distinction.

erictune commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,222 @@
+type ScheduledJobStatus struct {
+    Conditions []ScheduledJobCondition
+
+    // Successful tracks the overall amount of successful completions of this job.
```
This seems impractical to track since Jobs may eventually be garbage collected, and since, IIUC, status is supposed to be able to be recreated if it were deleted.

soltysh replied:
You've raised that issue already when implementing JobController. I'm thinking about taking different approach here. Instead of counting pods each time I'm thinking about updating this value with new information. Not sure how to implement it yet, though ;) Still the general idea of having those tracked should stay.

erictune commented on docs/proposals/scheduledjob.md:
```diff
@@ -0,0 +1,222 @@
+    ScheduledJobSuspended ScheduledJobConditionType = "Suspended"
+)
+
+// ScheduledJobCondition describes current state of a job.
```
If there is an overrun, then how do you have conditions for both?

Also, if this is just a copy of the JobStatus of the Job object, then an objectReference to the most recently created job would suffice?

soltysh replied:
You're right. I'll change the conditions array in status with an array of active job references.

soltysh replied:
This will also address the overrun problem, you've mentioned at the begining.

erictune commented:
This is getting close.  I think the main thing is to switch, as you suggest, from ISO8601, to another format (either cron or appengine format).  The other comments are pretty minor and needn't block you.

soltysh commented:
> AppEngine has a nice cron format that is similar to old fashioned cron, but more readable:
https://cloud.google.com/appengine/docs/python/config/cron?csw=1#Python_app_yaml_The_schedule_format

I think I'll stick with plain old cron ;)

soltysh commented:
I've rebased the changed to pick up all the verify changes. The last commit introduces following changes:
* change from ISO8601 to Cron schedule
* Epsilon -> StartingDeadlineSeconds
* ScheduledJobCondtion -> Active ObjectReferences array

@erictune ptal if this is ok, I can squash commits and if that's possible we could have it in before the end of this year :)

davidopp commented:
ping @erictune

bgrant0607 commented:
Google's cron system invokes a full-blown configuration system. Reasons:
* to update to the latest package (image) versions
* to instantiate multiple objects
* simple parameterization (e.g., to create a web of objects with unique names so multiple instances can exist simultaneously)

If one wanted to generate a workflow (#18827), or a Job with a PVC, or a Job with a ReplicationController for a pool of workers, there are 2 main alternatives:
1. Just launch a Job or Pod and use that Job/Pod to run a Kubernetes client that instantiates the desired resources
2. Create arbitrary objects via ScheduledJob

Problems with 1 include deletion of created resources in all failure scenarios and in the case of cancellation due to overlap.

One problem with both is propagating success or failure.

My personal preference is (2), because it removes a layer of indirection. At that point, it's more like a scheduled API call. The arbitrary template proposal is #18215.

bgrant0607 commented:
Regarding whether we need this or let people figure it out:
Thousands of teams in Google use something along these lines. We should make it easy.

xiang90 commented on docs/proposals/scheduledjob.md:
````diff
@@ -0,0 +1,197 @@
+}
+```
+
++### Running ScheduledJobs using kubectl
````
broken markdown. remove the extra `+` at the beginning.

soltysh commented:
Addressed last comment. 
@erictune mind taking a look?

erictune commented:
@bgrant0607 
Agree that we need templates at some point.  I think there is value in making progress on this and templates in parallel.  So, I suggest the following evolution path for ScheduledJob:

- Proceed with ScheduledJob as described here, except change `JobTemplate JobSpec` to be `JobTemplate *JobSpec`, thus allowing it to be optional in the future
- Later, when Workflow is completed, add `WorkflowTemplate *WorkflowSpec` as an alternative to `JobTemplate`.
- Even later, when arbitrary Templates are done, also add `GenericTemplate *Template` 
- Do not fret about it being called `ScheduledJob` even though it will be able to start things that are not Jobs.  It can run jobs and still evokes the right idea in a user.
- The spec may end up looking like this:

```
type ScheduledJobSpec {
....
// Specify exactly one of the following:
JobTemplate *JobSpec
WorkflowTemplate *WorkflowSpec 
GenericTemplate *Template
```

erictune commented:
@soltysh per previous comment, suggest change `JobTemplate JobSpec` to `JobTemplate *JobSpec`.

bgrant0607 commented:
@erictune Your proposal for OneOf several template types sounds fine.

Was it a deliberate decision to not add Job metadata to the Job template?

erictune commented:
No, that was not deliberate.  IIUC, Brian is saying that we might want to make the following substitution:
`JobTemplate *JobSpec` :arrow_right: `JobTemplate *JobTemplateSpec`, where `JobTemplateSpec` is a new type that looks like this:

```
type JobTemplateSpec struct {
        // Standard object's metadata.
        // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
        ObjectMeta `json:"metadata,omitempty"`

        // Specification of the desired behavior of the job.
        // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
        Spec JobSpec `json:"spec,omitempty"`
}
```

and likewise for WorkflowTemplate, should we ever get there.

I think we want to do it that way for consistency, and to allow users to name their created jobs.

erictune commented:
Okay, so summarizing steps to close out this PR:

- [ ] remove pointer in slice
- [ ] need pointer for `JobTemplate`
- [ ] change type of `JobTemplate` from `JobSpec` to `JobTemplateSpec`, and define that type
- [ ] squash
- [ ] LGTM

erictune commented:
@bgrant0607 wonder if you have an opinion on the `ScheduledJobSpec.UniqueLabelKey` approach in this PR, versus having the unique selector generation be part of Job itself.

bgrant0607 commented:
@erictune I'd prefer unique label/selector generation to be part of Job.

soltysh commented:
@erictune as requested:
- removed pointer from `Active`
- made `JobTemplate` pointer
- made `JobTemplate` be `JobTemplateSpec` type
- moved label generation to job (new section about `JobSpec` modifications)

I've also:
- fixed all the ISO8601 leftovers
- extended future evolution with proper links to workflow and templating

It's all squashed and waiting your sing-off.

erictune commented:
This looks great.   LGTM.

soltysh commented:
Thanks for your help!

alex-mohr commented:
@k8s-bot unit test this

soltysh commented:
@erictune any chance on having this one merged?

erictune commented:
Merging because this is just a new doc.  Build failure is known flaky test.

erictune commented:
@soltysh merged.

soltysh commented:
@erictune thank you!

yuvipanda commented:
Is this being implemented anywhere?

soltysh commented:
Not yet. I'm planning to start working on it after 1.2 is released.
On Feb 27, 2016 3:25 AM, "Yuvi Panda" <notifications@github.com> wrote:

> Is this being implemented anywhere?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/11980#issuecomment-189559468>
> .
>

haizaar commented:
[Jobs guide ](http://kubernetes.io/docs/user-guide/jobs/) says that "cron is expected in the next minor release". Do you plan this for 1.3 or 1.2.x?

erictune commented:
1.3

zacharynevin commented:
What is the projected release date for v1.3?

sdminonne commented:
@zacharynevin  see https://github.com/kubernetes/kubernetes/wiki/Release-1.3

untoldone commented:
Did this make it into v1.3 (https://github.com/kubernetes/kubernetes/releases/tag/v1.3.0) ? If yes, is this documented on http://kubernetes.io/ somewhere? I still see "Support for creating Jobs at specified times/dates (i.e. cron) is expected in 1.3."

wernight commented:
Looks like documentation on http://kubernetes.io/docs/user-guide/jobs/#future-work didn't get updated (still pointing here).
